utils/src/main/java/io/atomix/utils/serializer/BufferAwareByteArrayOutputStream.java

```diff
@@ -0,0 +1,17 @@
+package io.atomix.utils.serializer;
+
+import java.io.ByteArrayOutputStream;
+
+/**
+ * Exposes protected byte array length in {@link ByteArrayOutputStream}.
+ */
+final class BufferAwareByteArrayOutputStream extends ByteArrayOutputStream {
+
+    BufferAwareByteArrayOutputStream(int size) {
+        super(size);
+    }
+
+    int getBufferSize() {
```

Comment:
buf never shrinks, only grows, and flush resets the size so we need to check the size like this.